gnovm/pkg/gnolang/machine.go

```diff
@@ -1968,6 +1968,365 @@ func (m *Machine) PushFrameCall(cx *CallExpr, fv *FuncValue, recv TypedValue, is
+		//   - in alice.TopFuncFnNoncrossing:       (storage:[], current:caller, previous:nil)
+		//   - in alice.TopFuncFnNoncrossing.inner: (storage:[], current:caller, previous:nil)
+		bob.Do(alice.TopFuncFnNoncrossing())        // FAIL: caller != bob
+		// ----------------------------------------
+		// CASE rA3: function declared in func(cur realm, ...) (crossing) -- do not storage-cross
+		//   - in bob.Do:                           (storage:[], current:caller, previous:nil)
+		//   - alice.TopFuncFnCrossing:             (storage:[], current:alice,  previous:caller)
+		//   - alice.TopFuncFnCrossing.inner:       (storage:[], current:caller, previous:nil)
+		//     (cross doesn't matter in inner)
+		bob.Do(alice.TopFuncFnCrossing(cross))      // FAIL: caller != bob
+		//
+		// --------------------------------------------------------------------------------
+		// CASE rB1: method has nil receiver -- storage-cross to declared
+		//   - in bob.Do:                           (storage:[],    current:caller, previous:nil)
+		//   - in obj.Method:                       (storage:[bob], current:caller, previous:nil)
+		bob.Do((*bob.Object)(nil)).Method)          // SUCCESS: storage-crossed to bob
+		// ----------------------------------------
+		// CASE rB2: method has nil receiver -- storage-cross to (wrong) declared
+		//   - in bob.Do:                           (storage:[],      current:caller, previous:nil)
+		//   - in obj.Method:                       (storage:[alice], current:caller, previous:nil)
+		bob.Do((*alice.Object)(nil)).Method)        // FAIL: alice != bob
+		// ----------------------------------------
+		// CASE rB3: method has unattached receiver (inline) -- do not storage-cross
+		//   - in bob.Do:                           (storage:[], current:caller, previous:nil)
+		//   - in obj.Method:                       (storage:[], current:caller, previous:nil)
+		bob.Do((new(alice.Object).Method)           // FAIL: caller != bob
+		// ----------------------------------------
+		// CASE rB4: method has unattached receiver (var) -- do not storage-cross (same as rB3 above)
+		//   - in bob.Do:                           (storage:[], current:caller, previous:nil)
+		//   - in obj.Method:                       (storage:[], current:caller, previous:nil)
+		obj := new(alice.Object)
+		bob.Do(obj.Method)                          // FAIL: caller != bob
+		//
+		// --------------------------------------------------------------------------------
+		// CASE rC1: method has attached receiver -- storage-cross to receiver
+		//  - in bob.Do:                            (storage:[],    current:caller, previous:nil)
+		//  - in obj.Method:                        (storage:[bob], current:caller, previous:nil)
+		obj := new(alice.Object)
+		bob.SetObject(obj)                          // obj attached to bob
+		bob.Do(obj.Method)                          // SUCCESS: attached to bob
```

Comment:
if in` package alice`, `Method` is declared as:

```go
func (obj *Object) Method() { // CASE rB2~4, CASE rC1~5
        // mutate unintended exported global var by assignment 
        bob.Private2 = -1
        // mutate intended global var
		bob.PrivateFunc()
}
```

it can be inferred that a realm should never declare an `exported` global var, if there are any method/closure attached/injected, to avoid unintended mutation, right?

Reply:
... from an external realm. I think this might be the answer. 
I thought it might not be enough because you can still pass in objects like...

```go
func (obj *Object) Method(t *thing) {
        t.Private2 = -1
        t.PrivateFunc()
}
```

But if t is read-write (e.g. doesn't have readonly sticky bit set) then it's intended to be modifiable.
But 'bob' is a package, it was never passed in as a read-write arg, so it seems that a package is maybe the only special case we need to consider.